File: unit08-lab/namer.cpp
/* namer.cpp implements and tests class Name.
 *
 * Begun by: Dr. Jump, CS 315 at King's College
 * Completed by: Robert Lowman
 * Date: 11/3/2016
 */

#include <iostream>        // istream, ostream
#include <string>          // string
#include <cassert>         // assert()
using namespace std;


class Name
{
public:
  Name(const string &, const string &, const string &);
  string getFirst() const;
  string getMiddle() const;
  string getLast() const;
  string getFullName() const;
  void print(ostream &);
private:
  string myFirst,
         myMiddle,
         myLast;
};


/**************************************************
* Name constructs a name from three strings.      *
* Receive: first, middle and last, three strings. *
* Return: the triplet (first middle last).        *
***************************************************/
Name::Name(const string & first, const string & middle, const string & last)
   {
      myFirst = first;
      myMiddle = middle;
      myLast = last;
   }

/***************************************************
 * getFirst() extracts the first name of a Name object. *
 * Return: myFirst.                                *
 ***************************************************/
string Name::getFirst() const
   {
      return myFirst;
   }

/*****************************************************
 * getMiddle() extracts the middle name of a Name object. *
 * Return: myMiddle.                                 *
 *****************************************************/
string Name::getMiddle() const
   {
      return myMiddle;
   }

/**************************************************
 * getLast() extracts the last name of a Name object.  *
 * Return: myLast.                                *
 **************************************************/
string Name::getLast() const
   {
      return myLast;
   }

/***************************************************
 * getFullName() returns a full name in F-M-L order.    *
 * Return: myFirst, myMiddle, myLast.              *
 ***************************************************/
string Name::getFullName() const
   {
      return myFirst + ' ' + myMiddle + ' ' + myLast;
   }


/****************************************************
 * print() displays a Name object.                    *
 * Receive: out, the name of a stream.              *
 * Output: the strings in theName to buffer.        *
 * POST: out contains myFirst, myMiddle and myLast. *
 ****************************************************/
void Name::print(ostream & out)
   {
      out << getFullName();
   }



/******************************************
 * a simple driver to test our Name type. *
 ******************************************/
int main()
{
  Name aName("John", "Paul", "Jones");

  //assert( aName.getFirst() == "John" );
  //assert( aName.getMiddle() == "Paul" );
  //assert( aName.getLast() == "Jones" );
  //assert( aName.getFullName() == "John Paul Jones" );

  //aName.print(cout);  cout << endl;
  cout << "\n\nAll tests passed! " << endl;
}

File: unit08-hw/namer.rb
# namer.rb tests class Name and its operations
#
# Begun by: Dr. Jump, for CS 315 at King's College.
# Completed by: Robert Lowman
# Date: 11/3/2016
####################################################

require 'test/unit/assertions'   # needed for assert
include Test::Unit::Assertions

class Name

     def initialize(first, middle, last)
       @first, @middle, @last = first, middle, last
     end

     attr_accessor :first, :middle, :last

     def fullName()
       @first + " " + @middle + " " + @last
     end

     def print()
       puts fullName()
       return fullName()
     end

     def lfmi()
       return @last + ", " + @first + " " + @middle[0,1] + "."
     end
     
     def read()
       puts "Enter the name:\n"
       @first = gets
       @middle = gets
       @last = gets
     end
      
end


def testName
   name = Name.new("John", "Paul", "Jones")

   assert name.first == "John", "first failed"
   assert name.middle == "Paul", "middle failed"
   assert name.last == "Jones", "last failed"
   assert name.fullName == "John Paul Jones", "fullName failed"
   assert name.print == "John Paul Jones", "print failed"

   assert name.lfmi == "Jones, John P."

   name.first = "Robert"
   name.middle = "Craig"
   name.last = "Lowman"

   puts "Name after using mutators"
   assert name.print == "Robert Craig Lowman"

   name.read()

   print "Name read in: \n"
   puts name.fullName

   print "All tests passed!\n"
end

testName
